mm2src/coins/rpc_command/get_new_address.rs

```diff
@@ -0,0 +1,340 @@
+
+        let chain = params.chain.unwrap_or_else(|| hd_wallet.default_receiver_chain());
+        let gap_limit = params.gap_limit.unwrap_or_else(|| hd_wallet.gap_limit());
+
+        // Check if we can generate new address.
+        match can_get_new_address(coin, hd_wallet, &hd_account, chain, gap_limit)
+            .await?
+            .reason
+        {
+            Some(CantGetNewAddressReason::AddressLimitReached { max_addresses_number }) => {
+                return MmError::err(GetNewAddressRpcError::AddressLimitReached { max_addresses_number })
+            },
+            Some(CantGetNewAddressReason::EmptyAddressesLimitReached { gap_limit }) => {
+                return MmError::err(GetNewAddressRpcError::EmptyAddressesLimitReached { gap_limit })
+            },
+            // We can generate new address. Proceed.
+            None => (),
+        }
+
+        let HDAddress {
+            address,
+            derivation_path,
+            ..
+        } = coin
+            .generate_new_address(hd_wallet, hd_account.deref_mut(), chain)
+            .await?;
+        let balance = coin.known_address_balance(&address).await?;
+
+        Ok(GetNewAddressResponse {
+            new_address: HDAddressBalance {
+                address: address.to_string(),
+                derivation_path: RpcDerivationPath(derivation_path),
+                chain,
+                balance,
+            },
+        })
+    }
+
+    async fn can_get_new_address<Coin>(
+        coin: &Coin,
+        hd_wallet: &Coin::HDWallet,
+        hd_account: &Coin::HDAccount,
+        chain: Bip44Chain,
+        gap_limit: u32,
+    ) -> MmResult<CanGetNewAddressResponse, GetNewAddressRpcError>
+    where
+        Coin: HDWalletBalanceOps + Sync,
+        <Coin as HDWalletCoinOps>::Address: fmt::Display,
+    {
+        let known_addresses_number = hd_account.known_addresses_number(chain)?;
+        if known_addresses_number == 0 || gap_limit >= known_addresses_number {
+            return Ok(CanGetNewAddressResponse::allowed());
+        }
+
+        let max_addresses_number = hd_wallet.address_limit();
+        if known_addresses_number >= max_addresses_number {
+            return Ok(CanGetNewAddressResponse::not_allowed(
+                CantGetNewAddressReason::AddressLimitReached { max_addresses_number },
+            ));
+        }
+
+        let address_scanner = coin.produce_hd_address_scanner().await?;
+
+        // Address IDs start from 0, so the `last_known_address_id = known_addresses_number - 1`.
+        // At this point we are sure that `known_addresses_number > 0`.
+        let last_address_id = known_addresses_number - 1;
+
+        for address_id in (0..=last_address_id).rev() {
+            let empty_addresses_number = last_address_id - address_id;
+            if empty_addresses_number > gap_limit {
+                return Ok(CanGetNewAddressResponse::not_allowed(
```

Comment:
Will it maybe be a bit more straight-forward to explicitly return error here? I guess if `allowed == false` GUIs will have to display an error alert even in case of successful RPC response.

Reply:
It would make sense if the method was called differently.
Do you have ideas? @shamardy advised `is_new_address_allowed`. But anyway this RPC assumes that we return a boolean variable as a result.

Reply:
Well, `is_new_address_allowed` should also return bool as you mentioned, and I'm unsure for now what name it can have 🙂 The other questions is: is `can_get_new_address` required at all? `get_new_address` performs this check anyway, so it might be sufficient to have only this method.

Reply:
This is a good question. `air_dex` doesn't use `can_get_new_address` RPC at all, but I added it earlier for the purpose so that the GUI can disable the `Get new address` button.
> The other questions is: is can_get_new_address required at all? get_new_address performs this check anyway, so it might be sufficient to have only this method.

Probably, we could remove this RPC taking into account that this checking is implemented already, and returning the RPC wouldn't be difficult.

Reply:
> RPC at all, but I added it earlier for the purpose so that the GUI can disable the Get new address button.

They can keep internal state for this, disabling the button in case of getting `AddressLimitReached | EmptyAddressesLimitReached` from `get_new_address`. With this, it's likely that we won't have to return the `can_get_new_address` RPC 🙂

Reply:
I removed `can_get_new_address` RPC.